Update readme (#1)
DevExpress-Examples/XAF_CustomLogonParameters: README.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T1145991)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->
# XAF - Customize Logon Parameters

This example demonstrates how to customize standard authentication behavior and supply additional logon parameters in XAF applications. The default login form displays **User Name** and **Password** editors. This sample customizes the login form to show **Company** and **Application User** lookup editors instead of **User Name**.

![Login form](https://docs.devexpress.com/eXpressAppFramework/images/custom-logon-prameters-blazor.png)

## Implementation Details

### ASP.NET Core Blazor

> 1. Define a custom class to be added to the logon window. In this example, this is the [Company](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/Company.cs) class that contains company names. This class should contain a list of `ApplicationUser` objects as a part of a one-to-many relationship. Then, add the second part of this relationship to the [ApplicationUser](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/ApplicationUser.cs#L28) class generated by the Solution Wizard. And finally, add the `Company` class to your application’s [DbContext](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/EFCoreCustomLogonAllDbContext.cs#L43).

Reviewer: You start with "define a custom class to be added to the logon window". I'm not sure what that means. You already described the task above - "replace User Name with Company and Application User editors". We never said anything about "adding classes to the window". That might seem confusing, I think. I would rather remove that first sentence and go directly to the sentence about the Company class. It should be more straightforward, as we already mentioned that we need a Company editor. 

```suggestion
1. Define a [Company](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/Company.cs) class that contains company names. This class should contain a list of `ApplicationUser` objects as a part of a one-to-many relationship. Add the second part of this relationship to the [ApplicationUser](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/ApplicationUser.cs#L28) class generated by the Solution Wizard. Finally, add the `Company` class to your application’s [DbContext](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/EFCoreCustomLogonAllDbContext.cs#L43).
```

> 1. Define custom parameters. To do this, add a new class with parameters to your project. In this example, this is the [Company](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/Company.cs) class. This class should contain company names and a list of `ApplicationUser` objects as a part of a one-to-many relationship. Then, add the second part of this relationship to the [ApplicationUser](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/ApplicationUser.cs#L28) class generated by the Solution Wizard. And finally, add the `Company` class to your application’s [DbContext](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/EFCoreCustomLogonAllDbContext.cs#L43).

Reviewer 2: 'Define a custom class that we add to the logon window. In this example this is the Company class that contains company names. This class should contain a list of 'ApplicationUser'.....

Author: ```suggestion
1. Define a custom class to be added to the logon window. In this example, this is the [Company](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/Company.cs) class that contains company names. This class should contain a list of `ApplicationUser` objects as a part of a one-to-many relationship. Then, add the second part of this relationship to the [ApplicationUser](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/ApplicationUser.cs#L28) class generated by the Solution Wizard. And finally, add the `Company` class to your application’s [DbContext](./CS/EF/EFCoreCustomLogonAll.Module/BusinessObjects/EFCoreCustomLogonAllDbContext.cs#L43).
```

> 2. Add custom parameters to the login window. For an example, refer to the following file - [CustomLogonParameters.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomLogonParameters.cs).

Reviewer 2: "Create a custom logon parameter class that uses the Company class." (on this step we don't add anything to the login window)

Author: ```suggestion
2. Create a custom logon parameter class that uses the `Company` class. For an example, refer to the following file - [CustomLogonParameters.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomLogonParameters.cs).
```

> 2. Create a custom logon parameter class that uses the `Company` class. For an example, refer to the following file - [CustomLogonParameters.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomLogonParameters.cs).

Reviewer: For an example - looks very similar to "for example", so may be confusing. And it's not something that you would see often. 

```suggestion
2. Create a custom logon parameter class that uses the `Company` class. To see an example, refer to the following file - [CustomLogonParameters.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomLogonParameters.cs).
```

> 3. Implement [custom authentication](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthentication.cs).

Reviewer 2: ... that uses the created custom logon parameters.

Author: ```suggestion
3. Implement [custom authentication](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthentication.cs) that uses created custom logon parameters..
```

> 3. Implement [custom authentication](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthentication.cs) that uses created custom logon parameters..

Reviewer: ```suggestion
3. Implement [custom authentication](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthentication.cs) that uses custom logon parameters.
```

> 4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs).

Reviewer 2: ... that creates an instance of the created custom authentication

Author: ```suggestion
4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs) that creates an instance of the defined custom authentication.
```

> 4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs) that creates an instance of the created custom authentication.

Author: ```suggestion
4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs) that creates an instance of the defined custom authentication.
```

> 4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs) that creates an instance of the defined custom authentication.

Reviewer: Similar as in the sentence before. If you already say "custom", it already implies that it's something you "created" or "defined" manually in this example. You don't need to say "created custom" or "defined custom". Such phrases seem redundant. 

 ```suggestion
4. Implement a custom [authentication provider](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Security/CustomAuthenticationStandardProvider.cs) that creates an instance of the custom authentication strategy.
```

5. Pass custom classes to the Security System. For an example, refer to the following file - [Startup.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Startup.cs#L73-L78).

> 6. Add demo data.

Reviewer 2: link to https://github.com/DevExpress-Examples/xaf-custom-logon-parameters/blob/update-readme/CS/EF/EFCoreCustomLogonAll.Module/DatabaseUpdate/Updater.cs

Author: ```suggestion
6. Add [demo data](./CS/EF/EFCoreCustomLogonAll.Module/DatabaseUpdate/Updater.cs).
```

> 7. Generate a demo database.

Reviewer 2: link to https://docs.devexpress.com/eXpressAppFramework/404264/data-security-and-safety/security-system/authentication/customize-standard-authentication-behavior-and-supply-additional-logon-parameters/customize-authentication-behavior-blazor#generate-a-demo-database

Author: ```suggestion
7. Generate a [demo database](https://docs.devexpress.com/eXpressAppFramework/404264/data-security-and-safety/security-system/authentication/customize-standard-authentication-behavior-and-supply-additional-logon-parameters/customize-authentication-behavior-blazor#generate-a-demo-database).
```

> 5. Pass custom classes to the Security System. For an example, refer to the following file - [Startup.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Startup.cs#L73-L78).

Reviewer: ```suggestion
5. Pass custom classes to the Security System. To see an example, refer to the following file - [Startup.cs](./CS/EF/EFCoreCustomLogonAll.Blazor.Server/Startup.cs#L73-L78).
```
